File: app/src/main/java/com/xi/companion/MainActivity.java
package com.xi.companion;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    private static final int REQUEST_OVERLAY = 1001;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        checkOverlayPermission();
    }

    private void checkOverlayPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
            Toast.makeText(this, "请授予悬浮窗权限", Toast.LENGTH_LONG).show();
            Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                Uri.parse("package:" + getPackageName()));
            startActivityForResult(intent, REQUEST_OVERLAY);
        } else {
            startServiceAndFinish();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_OVERLAY) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && Settings.canDrawOverlays(this)) {
                startServiceAndFinish();
            } else {
                Toast.makeText(this, "权限被拒绝，应用无法运行", Toast.LENGTH_SHORT).show();
                finish();
            }
        }
    }

    private void startServiceAndFinish() {
        Intent serviceIntent = new Intent(this, XiOverlayService.class);
        startForegroundService(serviceIntent);
        Toast.makeText(this, "Xi Companion 服务已启动", Toast.LENGTH_SHORT).show();
        finish();
    }
}

File: app/src/main/java/com/xi/companion/MainActivity.kt
package com.xi.companion

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!Settings.canDrawOverlays(this)) {
            val intent = Intent(
                Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                Uri.parse("package:com.xi.companion")
            )
            startActivityForResult(intent, 123)
        } else {
            startOverlay()
        }
    }

    private fun startOverlay() {
        val intent = Intent(this, XiOverlayService::class.java)
        startService(intent)
        finish()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == 123 && Settings.canDrawOverlays(this)) {
            startOverlay()
        }
    }
}